tensorflow-cnn/cnn_example.R

```diff
@@ -0,0 +1,202 @@
+# Install packages
+install.packages("devtools", repos = "http://cran.us.r-project.org")
+library(devtools)
+install.packages("RgoogleMaps", repos = "http://cran.us.r-project.org")
+library(RgoogleMaps)
+install.packages("R.matlab", repos = "http://cran.us.r-project.org")
+library(R.matlab)
+install_github("sigopt/SigOptR")
+library(SigOptR)
+install_github("rstudio/tensorflow")
+library(tensorflow)
+
+Sys.setenv(SIGOPT_API_TOKEN="INSERT_YOUR_TOKEN_HERE")
+
+# load SVHN dataset
+extra_X <- R.matlab::readMat('extra_32x32.mat')$X
+extra_y <- R.matlab::readMat('extra_32x32.mat')['y']
+test_X <- R.matlab::readMat('test_32x32.mat')$X
+test_y <- R.matlab::readMat('test_32x32.mat')['y']
+train_X <- R.matlab::readMat('train_32x32.mat')$X
+train_y <- R.matlab::readMat('train_32x32.mat')['y']
+
+convert_rgb2gray <- function(X) {
+  X_gray <- array(0, c(32, 32, dim(X)[4]))
+  for (i in 1:dim(X)[4]) {
+    img_gray <- RGB2GRAY(X[,,,i])
+    X_gray[,,i] <- img_gray
+  }
+  return(X_gray)
+}
+
+# convert to grayscale
+extra_X <- convert_rgb2gray(extra_X)
+test_X <- convert_rgb2gray(test_X)
+train_X <- convert_rgb2gray(train_X)
+
+image_w <- 32
+extra_XZ <- array(aperm(extra_X), c(dim(extra_X)[3], image_w * image_w))
+test_XZ <- array(aperm(test_X), c(dim(test_X)[3], image_w * image_w))
+train_XZ <- array(aperm(train_X), c(dim(train_X)[3], image_w * image_w))
+
+# normalize image features
+extra_XZ <- t(scale(t(extra_XZ)))
+test_XZ <- t(scale(t(test_XZ)))
+train_XZ <- t(scale(t(train_XZ)))
+
+# one-hot encode labels
+extra_y$y <- factor(extra_y$y)
+extra_yZ <- model.matrix(~.-1, data = extra_y)
+test_y$y <- factor(test_y$y)
+test_yZ <- model.matrix(~.-1, data = test_y)
+train_y$y <- factor(train_y$y)
+train_yZ <- model.matrix(~.-1, data = train_y)
+
+# stack train, extra
+extra_XZ <- rbind(extra_XZ, train_XZ)
+extra_yZ <- rbind(extra_yZ, train_yZ)
+
+# take 75% of dataset
+index <- sample(1:nrow(extra_XZ), size=0.75 * nrow(extra_XZ))
+extra_XZ <- extra_XZ[index, ]
+index <- sample(1:nrow(extra_yZ), size=0.75 * nrow(extra_yZ))
+extra_yZ <- extra_yZ[index, ]
+
+# create SigOpt experiment
+experiment <- create_experiment(list(
+  name="SVHN ConvNet (R)",
+  parameters=list(
+    list(name="filter1_w", type="int", bounds=list(min=3, max=10)),
+    list(name="filter1_depth", type="int", bounds=list(min=10, max=64)),
+    list(name="filter2_w", type="int", bounds=list(min=3, max=10)),
+    list(name="filter2_depth", type="int", bounds=list(min=10, max=64)),
+    list(name="drp_out_keep_p", type="double", bounds=list(min=0.2, max=1.0)),
+    list(name="log_rms_lr", type="double", bounds=list(min=log(0.0001), max=log(1.0))),
+    list(name="rms_mom", type="double", bounds=list(min=0.5, max=1.0)),
+    list(name="rms_decay", type="double", bounds=list(min=0.5, max=1.0))
+  ),
+  observation_budget=100
+))
+
+# SigOpt optimization loop
```

Comment:
can we pull out the middle section of this loop into an `evaluate_metric` function? That way it's easier to see that the loop is:
 * create suggestion
 * evaluate metric
 * report observations
because the code will be so close together